junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/MutableExtensionRegistry.java

```diff
@@ -93,9 +95,23 @@ public static MutableExtensionRegistry createRegistryWithDefaultExtensions(Jupit
 		return extensionRegistry;
 	}
 
-	private static void registerAutoDetectedExtensions(MutableExtensionRegistry extensionRegistry) {
+	private static void registerAutoDetectedExtensions(MutableExtensionRegistry extensionRegistry,
+			JupiterConfiguration configuration) {
+		Predicate<String> filter = createExtensionFilterByPatterns(
+			configuration.getExtensionAutodetectionIncludePattern().orElse("*"),
+			configuration.getExtensionAutodetectionExcludePattern().orElse(""));
+
 		ServiceLoader.load(Extension.class, ClassLoaderUtils.getDefaultClassLoader())//
-				.forEach(extensionRegistry::registerAutoDetectedExtension);
+				.forEach(extension -> {
+					if (filter.test(extension.getClass().getName())) {
```

Comment:
I wonder whether it would be worth creating a Java 9+ variant that avoids instantiating extension that don't pass the filter. We could do this via a new `ServiceLoaderUtils` class in junit-platform-commons that has a Java 9+ implementation in `src/main/java9`. The signature should look sth. like

```java
Stream<T> load(
	Class<T> service,
	Predicate<? super Class<? extends T>> providerPredicate,
	ClassLoader loader
);
```

with the Java 8 implementation being

```java
StreamSupport.stream(ServiceLoader.load(service, loader).spliterator(), false)
	.filter(it -> {
		@SuppressWarnings("unchecked") Class<? extends T> type = (Class<? extends T>) it.getClass();
		return providerPredicate.test(type);
	});
```

and Java 9+

```java
ServiceLoader.load(service, loader).stream()
	.filter(provider -> providerPredicate.test(provider.type()))
	.map(ServiceLoader.Provider::get);
```

WDYT?

Reply:
@marcphilipp 

Sounds good! I’m on board with this approach! 👍🏻

Reply:
I implemented it as per the comments, but encountered the following exception.

- entry: META-INF/versions/9/org/junit/platform/commons/util/ServiceLoaderUtils.class, contains a class with different api from earlier version.

I would greatly appreciate it if you could point out any issues or mistakes in my implementation. 🥲

Reply:
@YongGoose That's because one of them is `final` while the other isn't. The other `*Utils` classes aren't final so I don't think they should be here, either. But adding a private constructor (to both implementations) sounds like a good idea.

Reply:
@marcphilipp 
Thank you!👍🏻 
As you suggested, removing the final keyword resolved the issue, and the tests passed successfully.